pkg/util/docker/event_stream.go

```diff
@@ -98,23 +123,59 @@ CONNECT:
 			case <-cancelChan:
 				cancel()
 				return
-			case msg := <-messages:
-				event, err := d.processContainerEvent(msg)
-				if err != nil {
-					log.Debugf("skipping event: %s", err)
-					continue
-				}
-				dataChan <- event
 			case err := <-errs:
 				if err == io.EOF {
 					// Silently ignore io.EOF that happens on http connection reset
-					log.Debug("got EOF, re-connecting")
+					log.Debug("Got EOF, re-connecting")
 				} else {
 					log.Warnf("error getting docker events: %s", err)
 				}
 				cancel()
 				continue CONNECT // Re-connect to docker
+			case msg := <-messages:
+				event, err := d.processContainerEvent(msg)
+				if err != nil {
+					log.Debugf("Skipping event: %s", err)
+					continue
+				}
+
+				badSubs := d.eventState.dispatch(event)
+				for _, sub := range badSubs {
+					d.UnsubscribeFromContainerEvents(sub.name)
+				}
 			}
 		}
 	}
 }
+
+func (e *eventStreamState) dispatch(event *ContainerEvent) []*eventSubscriber {
+	var badSubs []*eventSubscriber
+	e.RLock()
+	for _, sub := range e.subscribers {
+		err := sub.sendEvent(event)
+		if err != nil {
+			badSubs = append(badSubs, sub)
+			sub.sendError(err)
```

Comment:
I'm not sure about the `sendError`, we're sending an error to the subscriber because he returned us an error? 🤔